concepts/type-theory-glossary.md

```diff
@@ -0,0 +1,396 @@
+
+Sum types usually show up in prose descriptions of data as the word "or".
+
+For example, say: "I have a boolean *or* an int" --
+this would be an example of a sum type,
+and its cardinality would be "|inf+2|" --
+countable infinity for the int, and plus two for the boolean.
+
+A clearer example might take two enums -- let's call them
+"Animals" (it has members "cat", "dog", "rabbit"),
+and "Furniture" (it has members "chair", and "desk") --
+and say "I have an Animal *or* a furniture" --
+then clearly I have one (and only one) of these five things:
+the cardinality is "|3+2|".
+
+Sum types go by many names in various communities: they're also commonly
+known as "(discriminated) unions", "(tagged) variants", or other terminology.
+(We use the "union" term ourselves, in the lexicon of IPLD Schemas!)
+
+You can also consult the Wikipedia page on
+[Sum Types](https://en.wikipedia.org/wiki/Sum_types)
+for broader discussion of the topic.
+https://chadaustin.me/2015/07/sum-types/ also contains some useful content
+and examples from a variety of languages.
+
+### product types
+
+Product types have a cardinality that is *the multiplied cardinality of their member types*.
+
+This may sound outlandish, but you've been using them all your life:
+*"product types" are commonly known as "structs"*
+(or, "classes", in more aggressively object-oriented languages).
+
+Product types usually show up in prose descriptions of data as the word "and".
+
+For example, say: "I have a boolean and another boolean and another boolean" --
+you might write code for this as `struct{ a bool, b bool, c bool }`,
+and its cardinality would be "|2*2*2|" -- "|8|" --
+two for each of the booleans, and multiplied because each of them may vary independently,
+and thus each of the booleans adds another "dimension" to the state space described by this combined value.
+
+For another example, using our enums again above --
+"Animals" (it has members "cat", "dog", "rabbit"),
+and "Furniture" (it has members "chair", and "desk") --
+we could say "I have an Animal, *and* it's on top of a Furniture" --
+and then I could have any one of of *six* states
+("cat on a chair", "cat on desk", "dog on a chair", etc):
+the cardinality is "|3*2|".
+
+### infinite recursives
+
+Infinite recursives, as it says in the name, have infinite cardinality.
+
+Maps and lists are both examples of infinite recursives.
+Since we can keep inserting entries into a map, or appending values to a list,
+and still have a map or list respectively, we can't count their cardinality.
+
+Note that it's still useful to do math around these, though.
+For example, if you have "maybe a map" (see the [maybe types](#maybe-types) section,
+coming later, in the [Types in the Wild](#types-in-the-wild) chapter) you may
+find it useful to declare that the cardinality of your map is "a" (an arbitrary placeholder),
+and then the cardinality of "maybe map" is "|a+1|".
```

Comment:
For the mathematically inclined, it may be helpful to mention here that this is why we refer to "cardinality" rather than "ordinality" (since `order(a+1) = order(a)` when `a` is infinite).